File: evaluation/typescript/src/strategyEvaluator.ts
import { Bucketeer } from './bucketeer';
import { RolloutStrategy, Strategy } from './proto/feature/strategy_pb';
import { Variation } from './proto/feature/variation_pb';

class StrategyEvaluator {
  evaluate(
    strategy: Strategy,
    userID: string,
    variations: Variation[],
    featureID: string,
    samplingSeed: string,
  ): Variation {
    switch (strategy.getType()) {
      case Strategy.Type.FIXED:
        return this.findVariation(strategy.getFixedStrategy()?.getVariation() || '', variations);
      case Strategy.Type.ROLLOUT:
        const rolloutStrategy = strategy.getRolloutStrategy();
        if (rolloutStrategy !== undefined) {
          const variationID = this.rollout(rolloutStrategy, userID, featureID, samplingSeed);
          return this.findVariation(variationID, variations);
        }
        throw new Error('Missing rollout strategy');
      default:
        throw new Error('Unsupported strategy');
    }
  }

  private rollout(
    strategy: RolloutStrategy,
    userID: string,
    featureID: string,
    samplingSeed: string,
  ): string {
    const bucketeer = new Bucketeer();
    
    // Audience control: Check if user should be included in experiment
    const audience = strategy.getAudience();
    if (audience !== undefined) {
      if (audience.getPercentage() > 0 && audience.getPercentage() < 100) {
        // Use different hash input for audience control to ensure independence from A/B split
        const trafficInput = `traffic-${featureID}-${userID}-${samplingSeed}`;
        const trafficBucket = bucketeer.bucket(trafficInput);
        const trafficThreshold = audience.getPercentage() / 100.0;
        
        // If user is not in experiment traffic, return default variation
        if (trafficBucket >= trafficThreshold) {
          if (audience.getDefaultVariation() === '') {
            throw new Error('Variation not found');
          }
          return audience.getDefaultVariation();
        }
      }
    }
    
    // Original A/B split logic for users in experiment traffic
    const input = `${featureID}-${userID}-${samplingSeed}`;
    const bucket = bucketeer.bucket(input);

    let rangeEnd = 0.0;
    for (const variation of strategy.getVariationsList()) {
      rangeEnd += variation.getWeight() / 100000.0;
      if (bucket < rangeEnd) {
        return variation.getVariation();
      }
    }
    throw new Error('Variation not found');
  }

  private findVariation(variationID: string, variations: Variation[]): Variation {
    for (const variation of variations) {
      if (variation.getId() === variationID) {
        return variation;
      }
    }
    throw new Error('Variation not found');
  }
}

export { StrategyEvaluator };

File: evaluation/go/strategy_evaluator.go
// Copyright 2025 The Bucketeer Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package evaluation

import (
	"fmt"

	"github.com/bucketeer-io/bucketeer/proto/feature"
)

type strategyEvaluator struct {
}

func (e *strategyEvaluator) Evaluate(
	strategy *feature.Strategy,
	userID string,
	variations []*feature.Variation,
	featureID string,
	samplingSeed string,
) (*feature.Variation, error) {
	switch strategy.Type {
	case feature.Strategy_FIXED:
		return findVariation(strategy.FixedStrategy.Variation, variations)
	case feature.Strategy_ROLLOUT:
		variationID, err := e.rollout(strategy.RolloutStrategy, featureID, userID, samplingSeed)
		if err != nil {
			return nil, err
		}
		return findVariation(variationID, variations)
	}
	return nil, ErrUnsupportedStrategy
}

func (e *strategyEvaluator) rollout(
	strategy *feature.RolloutStrategy,
	featureID, userID, samplingSeed string,
) (string, error) {
	b := bucketeer{}
	bucket := b.bucket(fmt.Sprintf("%s-%s-%s", featureID, userID, samplingSeed))
	// Iterate through the variant and increment the threshold by the percentage of each variant.
	// return the first variant where the bucket is smaller than the threshold.
	rangeEnd := 0.0
	for i := range strategy.Variations {
		rangeEnd += float64(strategy.Variations[i].Weight) / 100000.0
		if bucket < rangeEnd {
			return strategy.Variations[i].Variation, nil
		}
	}
	return "", ErrVariationNotFound
}
